File: src/python/grid.py
import numpy as np
from molecule import Molecule
from scipy.integrate import simpson

class Grid:
	def __init__(self):
		self.scfp_values = list()

	def calculate_scfp_values(self, molecule: Molecule):
		if (not molecule.density_matrix.size == 0) or (not len(molecule.orbitals) == 0):
			for x, y, z in zip(self.x, self.y, self.z):
				self.scfp_values.append(molecule.scfp_density_at_point((x,y,z)))

			self.scfp_values = np.array(self.scfp_values)
			self.scfp_values_normalized = self.scfp_values/self.scfp_values.max()
			self.grid_points_sizes = self.scfp_values_normalized*20
		else:
			raise Exception("The molecule has niether density matrix nor orbitals")
		return self.scfp_values


	def draw_grid_points(self, ax, type=None, molecule=None, number=None):
		if type == "scfp":
			if len(self.scfp_values):
				ax.scatter(self.x, self.y, self.z, cmap="Blues", c=self.scfp_values, s=self.grid_points_sizes, alpha=0.4)
			else:
				raise Exception("You should first calculate scfp values using calculate_scfp_values() function")
		elif type == "orbital":
			if number == None:
				raise Exception("You should specify the number of orbital")

			values = list()
			for x, y, z in zip(self.x, self.y, self.z):
				values.append(molecule.orbital_value_at_point((x,y,z), number))

			ax.scatter(self.x, self.y, self.z, cmap="PiYG", c=values, s=7, alpha=0.4)
		else:
			ax.scatter(self.x, self.y, self.z, c="blue", s=5, alpha=0.4)

	def draw_isosurface(self, ax, value):
		x = np.asarray(self.x)
		y = np.asarray(self.y)
		z = np.asarray(self.z)

		tol = 0.01
		mask = self.scfp_values >= value

		ax.scatter(self.x[mask], self.y[mask], self.z[mask], c="red", s=15, alpha=0.5)

	def scfp_integral(self):
		raise NotImplementedError("Subclass must implement abstract method")


class SphericalGrid(Grid):
	def __init__(self, phi_points=20, theta_points=10, r_points=10, radius=3, center=(0,0,0)):
		super().__init__()
		self.delta_phi = 2*np.pi/phi_points
		phi_initial = self.delta_phi/2
		phi_final = 2*np.pi-(self.delta_phi/2)

		self.delta_theta = np.pi/theta_points
		theta_initial = self.delta_theta/2
		theta_final = np.pi-(self.delta_theta/2)

		self.delta_r = 2*radius/(2*r_points-1)
		r_initial = self.delta_r/2

		phi = np.linspace(phi_initial, phi_final, phi_points)
		theta = np.linspace(theta_initial, theta_final, theta_points)
		r = np.linspace(r_initial, radius, r_points)

		r, theta, phi = np.meshgrid(r, theta, phi) 

		# Convert to Cartesian coordinates
		x = center[0] + r * np.sin(theta) * np.cos(phi)
		y = center[1] + r * np.sin(theta) * np.sin(phi)
		z = center[2] + r * np.cos(theta)

		self.x = x.flatten()
		self.y = y.flatten()
		self.z = z.flatten()

		self.r = r.flatten()
		self.phi = phi.flatten()
		self.theta = theta.flatten()

		self.center_point = center

	
class RegularOrthogonalGrid(Grid):
	def __init__(self, x_length=4, y_length=4, z_length=4, x_points=10, y_points=10, z_points=10, center=(0,0,0)):
		super().__init__()
		self.x_points = np.linspace(center[0] - x_length/2, center[0] + x_length/2, x_points)
		self.y_points = np.linspace(center[1] - y_length/2, center[1] + y_length/2, y_points)
		self.z_points = np.linspace(center[2] - z_length/2, center[2] + z_length/2, z_points)

		x, y, z = np.meshgrid(self.x_points, self.y_points, self.z_points)

		self.x = x.flatten()
		self.y = y.flatten()
		self.z = z.flatten()

		self.delta_x = x_length/(x_points-1)
		self.delta_y = y_length/(y_points-1)
		self.delta_z = z_length/(z_points-1)
		self.delta_V = self.delta_x * self.delta_y * self.delta_z

		self.center_point = center

	@classmethod
	def from_molecule(cls, molecule: Molecule, x_points=20, y_points=20, z_points=20):
		x_margin = 2  # value in Angstroms
		y_margin = 2 
		z_margin = 2

		x_length = molecule.x_max - molecule.x_min + x_margin
		y_length = molecule.y_max - molecule.y_min + y_margin
		z_length = molecule.z_max - molecule.z_min + z_margin

		center_point = ((molecule.x_min+molecule.x_max)/2, (molecule.y_min+molecule.y_max)/2, (molecule.z_min+molecule.z_max)/2)

		return cls(x_length, y_length, z_length, x_points, y_points, z_points, center_point)

	def scfp_integral(self):
		# return np.sum(self.scfp_values) * self.delta_V
		Nx, Ny, Nz = len(self.x_points), len(self.y_points), len(self.z_points)
		f = self.scfp_values.reshape((Nx, Ny, Nz))
		I = simpson(
		        simpson(
		            simpson(f, self.z_points, axis=2),
		            self.y_points, axis=1),
		        self.x_points, axis=0
		    )
		return I


class IrregularOrthogonalGrid(Grid):
	def __init__(self, delta=0.01, scaling_factor=1.3, points=20, center=(0,0,0)):
		super().__init__()
		template = np.cumsum(((scaling_factor**np.linspace(0, points-1, points))*delta))
		print(template)

		self.x_points = np.append(center[0] + template, center[0] - template)
		self.y_points = np.append(center[1] + template, center[1] - template)
		self.z_points = np.append(center[2] + template, center[2] - template)

		self.x_points = np.append(center[0], self.x_points)
		self.y_points = np.append(center[1], self.y_points)
		self.z_points = np.append(center[2], self.z_points)
		
		print(self.x_points)

		self.x_points = np.sort(self.x_points)
		self.y_points = np.sort(self.y_points)
		self.z_points = np.sort(self.z_points)

		print(self.x_points)


		x, y, z = np.meshgrid(self.x_points, self.y_points, self.z_points)

		self.x = x.flatten()
		self.y = y.flatten()
		self.z = z.flatten()

		print(self.x)
		print(self.y)
		print(self.z)

		self.center_point = center

	def scfp_integral(self):
		# return np.sum(self.scfp_values) * self.delta_V
		Nx, Ny, Nz = len(self.x_points), len(self.y_points), len(self.z_points)
		f = self.scfp_values.reshape((Nx, Ny, Nz))
		I = simpson(
		        simpson(
		            simpson(f, self.z_points, axis=2),
		            self.y_points, axis=1),
		        self.x_points, axis=0
		    )
		return I

File: src/cpp/Grid.cpp
#include "Grid.h"

// -------- SphericalGrid --------

SphericalGrid::SphericalGrid(int phi_points, int theta_points, int r_points,
                             double radius, std::array<double, 3> center)
    : delta_phi(2 * M_PI / phi_points),
      delta_theta(M_PI / theta_points),
      delta_r(2 * radius / (2 * r_points - 1))
{
    center_point_ = center;

    double phi_initial = delta_phi / 2.0;
    double phi_final = 2 * M_PI - (delta_phi / 2.0);
    double theta_initial = delta_theta / 2.0;
    double theta_final = M_PI - (delta_theta / 2.0);
    double r_initial = delta_r / 2.0;

    std::vector<double> phi(phi_points), theta(theta_points), r(r_points);
    for (int i = 0; i < phi_points; ++i)
        phi[i] = phi_initial + i * (phi_final - phi_initial) / (phi_points - 1);
    for (int i = 0; i < theta_points; ++i)
        theta[i] = theta_initial + i * (theta_final - theta_initial) / (theta_points - 1);
    for (int i = 0; i < r_points; ++i)
        r[i] = r_initial + i * (radius - r_initial) / (r_points - 1);

    for (double rv : r) {
        for (double tv : theta) {
            for (double pv : phi) {
                double X = center[0] + rv * std::sin(tv) * std::cos(pv);
                double Y = center[1] + rv * std::sin(tv) * std::sin(pv);
                double Z = center[2] + rv * std::cos(tv);
                x_.push_back(X);
                y_.push_back(Y);
                z_.push_back(Z);
                this->r.push_back(rv);
                this->theta.push_back(tv);
                this->phi.push_back(pv);
            }
        }
    }
}

// -------- RegularOrthogonalGrid --------

RegularOrthogonalGrid::RegularOrthogonalGrid(double x_length, double y_length, double z_length,
                                             int x_points_count, int y_points_count, int z_points_count,
                                             std::array<double, 3> center)
                                                 : delta_x_(0), delta_y_(0), delta_z_(0), delta_V_(0)
{
    center_point_ = center;
    number_of_x_points_ = x_points_count;
    number_of_y_points_ = y_points_count;
    number_of_z_points_ = z_points_count;
    create_grid(x_length, y_length, z_length, x_points_count, y_points_count, z_points_count, center);
}

RegularOrthogonalGrid::RegularOrthogonalGrid(const Molecule& molecule,
                                             int x_points_count, int y_points_count, int z_points_count)
                                                 : delta_x_(0), delta_y_(0), delta_z_(0), delta_V_(0)
{
    center_point_ = {
        (molecule.get_x_min() + molecule.get_x_max()) / 2.0,
        (molecule.get_y_min() + molecule.get_y_max()) / 2.0,
        (molecule.get_z_min() + molecule.get_z_max()) / 2.0
    };
    
    double x_margin = 18.0;
    double y_margin = 18.0;
    double z_margin = 18.0;

    number_of_x_points_ = x_points_count;
    number_of_y_points_ = y_points_count;
    number_of_z_points_ = z_points_count;

    double x_length = molecule.get_x_max() - molecule.get_x_min() + x_margin;
    double y_length = molecule.get_y_max() - molecule.get_y_min() + y_margin;
    double z_length = molecule.get_z_max() - molecule.get_z_min() + z_margin;

    double x_points_per_bohr = x_points_count / x_length;
    double y_points_per_bohr = y_points_count / y_length;
    double z_points_per_bohr = z_points_count / z_length;

    std::cout << "Grid. Points per 1 Bohr (x, y, x): (" << x_points_per_bohr << ", " << y_points_per_bohr << ", " << z_points_per_bohr << ")\n";

    create_grid(x_length, y_length, z_length, x_points_count, y_points_count, z_points_count, center_point_);
}

void RegularOrthogonalGrid::create_grid(double x_length, double y_length, double z_length,
                                        int x_points_count, int y_points_count, int z_points_count,
                                        std::array<double, 3> center)
{
    std::cout << center[0] << " " << center[1] << " " << center[2] << std::endl;
    x_points_.resize(x_points_count);
    y_points_.resize(y_points_count);
    z_points_.resize(z_points_count);

    delta_x_ = x_length / (x_points_count - 1);
    delta_y_ = y_length / (y_points_count - 1);
    delta_z_ = z_length / (z_points_count - 1);

    double x_min = center[0] - x_length / 2.0;
    double y_min = center[1] - y_length / 2.0;
    double z_min = center[2] - z_length / 2.0;

    for (int i = 0; i < x_points_count; ++i)
        x_points_[i] = x_min + i * delta_x_;
    for (int i = 0; i < y_points_count; ++i)
        y_points_[i] = y_min + i * delta_y_;
    for (int i = 0; i < z_points_count; ++i)
        z_points_[i] = z_min + i * delta_z_;

    for (double xv : x_points_)
        for (double yv : y_points_)
            for (double zv : z_points_) {
                x_.push_back(xv);
                y_.push_back(yv);
                z_.push_back(zv);
            }

    delta_V_ = delta_x_ * delta_y_ * delta_z_;
}


// -------- IrregularOrthogonalGrid --------

IrregularOrthogonalGrid::IrregularOrthogonalGrid(double initialDelta, double scaling_factor,
                                                 int numberOfPoints, std::array<double, 3> center)
{
    center_point_ = center;

    createGrid(initialDelta, scaling_factor, numberOfPoints, center);
}

IrregularOrthogonalGrid::IrregularOrthogonalGrid(double initialDelta, int numberOfPoints, std::array<double, 3> center,
    double xMax)
{
    double A = xMax / initialDelta;
    double alpha0 = 1;
    double N = numberOfPoints;
    // double scaling_factor = (std::log(A * (alpha0 + 1) / N) - alpha0 / (alpha0 + 1)) / (std::log(N) - 1 / (alpha0 + 1));

    for (int i = 0; i < 3; ++i) {
        alpha0 = (std::log(A * (alpha0 + 1) / N) - alpha0 / (alpha0 + 1)) / (std::log(N) - 1 / (alpha0 + 1));
        std::cout << "Step " << i << ": scaling_factor = " << alpha0 << "\n";
    }


    createGrid(initialDelta, alpha0, numberOfPoints, center);
}

void IrregularOrthogonalGrid::createGrid(double initialDelta, double scaling_factor, int numberOfPoints,
    std::array<double, 3> center)
{
    std::vector<double> template_vals(numberOfPoints);
    template_vals[0] = initialDelta;
    for (int i = 1; i < numberOfPoints; ++i)
        template_vals[i] = template_vals[i - 1] + std::pow(i, scaling_factor) * initialDelta;

    std::cout << "Parameters of the irregular grid:\n";
    std::cout << "min delta = " << initialDelta << "; max delta = " << template_vals[numberOfPoints-1] - template_vals[numberOfPoints-2] << std::endl;

    std::vector<double> positive(template_vals.begin(), template_vals.end());
    std::vector<double> negative(template_vals.begin(), template_vals.end());
    std::reverse(negative.begin(), negative.end());
    for (double& v : negative) v *= -1;

    x_points_.reserve(positive.size() + negative.size());
    y_points_.reserve(x_points_.size());
    z_points_.reserve(x_points_.size());

    // x_points_.push_back(center[0]);
    // y_points_.push_back(center[1]);
    // z_points_.push_back(center[2]);

    for (double v : negative) {
        x_points_.push_back(center[0] + v + initialDelta/2);
        y_points_.push_back(center[1] + v + initialDelta/2);
        z_points_.push_back(center[2] + v + initialDelta/2);
    }
    for (double v : positive) {
        x_points_.push_back(center[0] + v - initialDelta/2);
        y_points_.push_back(center[1] + v - initialDelta/2);
        z_points_.push_back(center[2] + v - initialDelta/2);
    }

    std::sort(x_points_.begin(), x_points_.end());
    std::sort(y_points_.begin(), y_points_.end());
    std::sort(z_points_.begin(), z_points_.end());

    std::cout << "min x = " << *std::min_element(x_points_.begin(), x_points_.end()) \
        << "; max x = " << *std::max_element(x_points_.begin(), x_points_.end()) << std::endl;

    for (double xv : x_points_)
        for (double yv : y_points_)
            for (double zv : z_points_) {
                x_.push_back(xv);
                y_.push_back(yv);
                z_.push_back(zv);
            }
}
